---
layout: issue
title: "LDAP connection bug with special characters"
id: ZF-10687
---

ZF-10687: LDAP connection bug with special characters
-----------------------------------------------------

 Issue Type: Bug Created: 2010-11-16T00:05:51.000+0000 Last Updated: 2012-01-05T09:44:26.000+0000 Status: Closed Fix version(s): 
 Reporter:  dolphy phan (dolphy)  Assignee:  Stefan Gehrig (sgehrig)  Tags: - Zend\_Ldap
 
 Related issues: 
 Attachments: 
### Description

authentication against Microsoft ADS (LDAP) does not work with special characters inside username! (latest version support only "-" and "\_" If we using a DN Form of authentication ex. "user.name@abc.com" seem does not work for usernames with special characters! Hope, that this fix will find his way in one of the next releases, because as a workaround

Regards, Dolphy

 

 

### Comments

Posted by Stefan Gehrig (sgehrig) on 2010-11-16T10:04:35.000+0000

Could you please provide some more information? Especially I'd need the configuration you use and some code snippet that shows your code setup.

I'm currently quite unsure where that limitation might come from.

 

 

Posted by dolphy phan (dolphy) on 2010-11-16T18:50:38.000+0000

Here is my config:

alwaysUseDatabaseForLogin = "admin" authAdapterMethod = "ldap" ldap.options.host = "192.168.199.2" ldap.options.port = "3268" ldap.options.username = "" ldap.options.password = "" ldap.options.useStartTls = "0" ldap.options.useSsl = "0" ldap.options.bindRequiresDn = "0" ldap.options.accountCanonicalForm = "4" ldap.options.allowEmptyPassword = "0" ldap.options.optReferrals = "0" ldap.options.baseDn = "cn=%s,dc=pyco,dc=vn" ldap.options.bindDnFormat = "%s@pyco.vn" ldap.options.accountDomainName = "pyco.vn" ldap.options.accountDomainNameShort = "pyco" ldap.options.accountFilterFormat = "" ldap.createGroupForUserInDB = "0" ldap.createUserInDBOnLogin = "0"

 

 

Posted by Andreas Heigl (heiglandreas) on 2010-11-17T01:11:20.000+0000

What kind of special characters are we talking about? Are there some apostrophes or Umlauts? or do you simply refer to a slash as special character?

And if you could provide some code snippets and examples of usernames that fail and that work it would be most helpful.

 

 

Posted by dolphy phan (dolphy) on 2010-11-21T18:42:21.000+0000

authentication ex. "user.name@abc.com" < dot inside username will failed to login.

 

 

Posted by dolphy phan (dolphy) on 2010-11-21T22:38:22.000+0000

Here is the errors I got
------------------------

An unexpected failure occurred 0x1 (Operations error; 00000000: LdapErr: DSID-0C090627, comment: In order to perform this operation a successful bind must be completed on the connection., data 0, vece): searching: (&(objectClass=user)(sAMAccountName=khoa.phan)) host=192.168.199.2,port=3268,username=,password=\*\*\*\*\*,useStartTls=0,useSsl=0,bindRequiresDn=0,accountCanonicalForm=4,allowEmptyPassword=0,optReferrals=0,baseDn=cn=%s,dc=pyco,dc=vn,accountDomainName=pyco.vn,accountDomainNameShort=pyco,accountFilterFormat= /var/www/html/usvndev/library/Zend/Ldap.php(987): 0x1 (Operations error; 00000000: LdapErr: DSID-0C090627, comment: In order to perform this operation a successful bind must be completed on the connection., data 0, vece): searching: (&(objectClass=user)(sAMAccountName=khoa.phan))

0 /var/www/html/usvndev/library/Zend/Ldap.php(666): Zend\_Ldap->search('(&(objectClass=...', 'cn=%s,dc=pyco,d...', 1, Array) #1 /var/www/html/usvndev/library/Zend/Ldap.php(538): Zend\_Ldap->\_getAccount('khoa.phan', Array) #2 /var/www/html/usvndev/library/Zend/Ldap.php(605): Zend\_Ldap->\_getAccountDn('khoa.phan@pyco....') #3 /var/www/html/usvndev/library/Zend/Auth/Adapter/Ldap.php(329): Zend\_Ldap->getCanonicalAccountName('khoa.phan@pyco....', 1) #4 /var/www/html/usvndev/library/Zend/Auth.php(117): Zend\_Auth\_Adapter\_Ldap->authenticate() #5 /var/www/html/usvndev/app/controllers/LoginController.php(73): Zend\_Auth->authenticate(Object(USVN\_Auth\_Adapter\_Ldap)) #6 /var/www/html/usvndev/app/controllers/LoginController.php(32): LoginController->\_doLogin() #7 /var/www/html/usvndev/library/Zend/Controller/Action.php(513): LoginController->loginAction() #8 /var/www/html/usvndev/library/Zend/Controller/Dispatcher/Standard.php(295): Zend\_Controller\_Action->dispatch('loginAction') #9 /var/www/html/usvndev/library/Zend/Controller/Front.php(954): Zend\_Controller\_Dispatcher\_Standard->dispatch(Object(USVN\_Controller\_Request\_Http), Object(Zend\_Controller\_Response\_Http)) #10 /var/www/html/usvndev/public/index.php(5): Zend\_Controller\_Front->dispatch() #11 {main}
====================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================

khoa.phan@pyco.vn authentication failed: 0x1 (Operations error; 00000000: LdapErr: DSID-0C090627, comment: In order to perform this operation a successful bind must be completed on the connection., data 0, vece): searching: (&(objectClass=user)(sAMAccountName=khoa.phan))

 

 

Posted by Stefan Gehrig (sgehrig) on 2010-11-22T10:41:41.000+0000

Looks strange, and I actually doubt that this error is related to the "dotted" username. Could you please try and remove the following lines from your configuration?

 
    <pre class="highlight">
    ldap.options.username = ""
    ldap.options.password = ""


 

 

Posted by dolphy phan (dolphy) on 2010-11-24T18:01:10.000+0000

Hi Gehriq,

I remove those two lines in my config and got this errors

Account not found: khoa.phan@pyco.vn 0x20 (No such object; 0000208D: NameErr: DSID-031001CD, problem 2001 (NO\_OBJECT), data 0, best match of: 'DC=pyco,DC=vn' ): searching: (&(objectClass=user)(sAMAccountName=khoa.phan)) host=192.168.199.2,port=3268,useStartTls=0,useSsl=0,bindRequiresDn=0,accountCanonicalForm=4,allowEmptyPassword=0,optReferrals=0,baseDn=cn=%s,dc=pyco,dc=vn,accountDomainName=pyco.vn,accountDomainNameShort=pyco,accountFilterFormat= khoa.phan@pyco.vn authentication failed: 0x20 (No such object; 0000208D: NameErr: DSID-031001CD, problem 2001 (NO\_OBJECT), data 0, best match of: 'DC=pyco,DC=vn' ): searching: (&(objectClass=user)(sAMAccountName=khoa.phan))

Ofcourse this account is avaialble.

 

 

Posted by dolphy phan (dolphy) on 2010-11-24T18:32:47.000+0000

Hi again,

I tried to change config again n got this error ---------config.ini------------ alwaysUseDatabaseForLogin = "admin" authAdapterMethod = "ldap" ldap.options.host = "192.168.199.2" ldap.options.port = "3268" ldap.options.useStartTls = "0" ldap.options.useSsl = "0" ldap.options.bindRequiresDn = "0" ldap.options.accountCanonicalForm = "4" ldap.options.allowEmptyPassword = "0" ldap.options.optReferrals = "0" ldap.options.baseDn = "dc=pyco,dc=vn" ldap.options.bindDnFormat = "%s@pyco.vn" ldap.options.accountDomainName = "pyco.vn" ldap.options.accountDomainNameShort = "pyco" ldap.options.accountFilterFormat = "" ldap.createGroupForUserInDB = "0" ldap.createUserInDBOnLogin = "0" -----------/config.ini--------------

Here the error

---------------error------------- [Thu Nov 25 09:24:58 2010] [error] [client 192.168.10.11] PHP Notice: Zend\_Loader::Zend\_Loader::registerAutoload is deprecated as of 1.8.0 and will be removed with 2.0.0; use Zend\_Loader\_Autoloader instead in /var/www/html/usvndev/library/Zend/Loader.php on line 266, referer: <http://192.168.199.1/udev/login/> --------------/error------------

 

 

Posted by Stefan Gehrig (sgehrig) on 2010-11-24T23:57:46.000+0000

OK - that looks completely different now... Zend\_Ldap does not have any dependency on Zend\_Loader and so does Zend\_Auth\_Adapter\_Ldap. The last warning (a PHP user notice to be precise) must be originated from your own user-land code though.

 

 

Posted by dolphy phan (dolphy) on 2010-11-26T01:36:49.000+0000

Hi,

Yes, I got login successfully with ldap config... but I got this.

--------------------------ERROS---------------------------- Invalid login. Logins can only include alphanumeric characters,dashes ('-'), or underscores ('\_')

Trace:

/var/www/html/usvndev/library/USVN/Db/Table/Row/User.php:184 USVN\_Db\_Table\_Row\_User::checkLogin  
 /var/www/html/usvndev/library/Zend/Db/Table/Row/Abstract.php:460 USVN\_Db\_Table\_Row\_User::\_insert /var/www/html/usvndev/library/Zend/Db/Table/Row/Abstract.php:437 Zend\_Db\_Table\_Row\_Abstract::\_doInsert  
 /var/www/html/usvndev/library/USVN/User.php:92 Zend\_Db\_Table\_Row\_Abstract::save  
 /var/www/html/usvndev/app/controllers/LoginController.php:123 USVN\_User::save /var/www/html/usvndev/app/controllers/LoginController.php:32 LoginController::\_doLogin  
 /var/www/html/usvndev/library/Zend/Controller/Action.php:513 LoginController::loginAction  
 /var/www/html/usvndev/library/Zend/Controller/Dispatcher/Standard.php:289 Zend\_Controller\_Action::dispatch  
 /var/www/html/usvndev/library/Zend/Controller/Front.php:946 Zend\_Controller\_Dispatcher\_Standard::dispatch  
 /var/www/html/usvndev/public/index.php:5 Zend\_Controller\_Front::dispatch --------------------------ERROS----------------------------

 

 

Posted by Stefan Gehrig (sgehrig) on 2010-11-26T01:59:58.000+0000

Again - nothing to do with Zend\_Ldap:

 
    <pre class="highlight">
    USVN_Db_Table_Row_User::checkLogin()


is the source of your error message.

 

 

Posted by Stefan Gehrig (sgehrig) on 2010-11-26T02:00:40.000+0000

Obviously not a Zend\_Ldap issue.

 

 

Posted by devaud (devaud) on 2012-01-05T09:44:26.000+0000

The issue is: go to this file: /var/www/html/usvndev/library/USVN/Db/Table/Row/User.php (or this /var/www/usvn/library/USVN/Db/Table/Row/User.php for me) edit it In the "check login" row, you just have to add a point into the code:

if (!preg\_match('/^[0-9a-zA-Z\_.-]+$/', $login)) {

I dunno why the developpers didnt add it, but i just know that it works with that. Hope it helps

 

 